Fix testLineAgainstWater to work in all circumstances

forkerer opened the pull request:
This pull request aims to fix #564, it adds new way of detecting line/water collision, i built it as a part of CWater* classes.
New mode also works outside of GTA SA original map boundaries, while old one returns false every time there.

Easiest way to test this is to check water detection here:

![image](https://user-images.githubusercontent.com/923242/45929640-7ff54400-bf4c-11e8-882c-7ca74e9ee38f.png)

or outside of map boundaries.

New function looks like this:
```lua
bool float float float testLineAgainstWater ( float startX, float startY, float startZ, float endX, float endY, float endZ )
```

Test resource:
[water.zip](https://github.com/multitheftauto/mtasa-blue/files/2408806/water.zip)

Also fixes #490, see https://github.com/multitheftauto/mtasa-blue/pull/568#issuecomment-423830046

EDIT:
*** Legacy mode removed

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Added more precise testLineAgainstWater

forkerer commented:
Also, this fixes https://github.com/multitheftauto/mtasa-blue/issues/490 too.

![image](https://user-images.githubusercontent.com/923242/50720906-61a3f600-10ae-11e9-8456-4698d3291efa.png)

Einheit-101 commented:
#490 and this one is exactly one and the same bug report and this fix comes in handy for my server

qaisjp commented:
Looks good. Is there any use in keeping legacy mode at all?

AlexTMjugador commented:
Backward compatibility could be seen as one use. It might happen that some obscure (and compiled) script relies on the behavior documented in #490 for its purposes, but I find that possibility so improbable. It's not like that has any useful purpose I can think of.

forkerer commented:
> Looks good. Is there any use in keeping legacy mode at all?

Backward compatibility was my main reasoning, but i gotta agree with AlexTMjugador that it seems unlikely that it would affect any of the currently existing scripts.

qaisjp commented:
I'm happy to approve this once it's made default (with no legacy mode)!

Commit:
Merge remote-tracking branch 'multitheftauto/master' into TestLineAgainstWater

Commit:
Removed legacy mode

Commit:
Fix remaining issues

- In some rare circumstances, array out of bound access was possible, resulting in read access exception.
- added early out in case of checked segment being out of map bounds, and not intersecting with water surface.
- fixed issue where checked segment was extending infinitely in reverse direction when checking for water out of map bounds.

forkerer commented:
@qaisjp removed legacy mode, also fixed some newly found issues. Tested it pretty extensively now, and can't find any cases where it's not working as intended.

qaisjp commented on Client/game_sa/CWaterManagerSA.cpp:
```diff
@@ -738,9 +803,89 @@ void CWaterManagerSA::SetWaveLevel(float fWaveLevel)
+    std::vector<CWaterZoneSA*> vecZones;
+    GetZonesIntersecting(vecStart, vecEnd, vecZones);
+
+    if (vecZones.size() < 1)
```
```suggestion
    if (vecZones.empty())
```

qaisjp reviewed (approved):
Tiny suggestion, otherwise looks good.
